File: 2022/Sources/advent2022/Day3.swift
struct Day3 : Day {

    let sampleInput = [
        "vJrwpWtwJgWrhcsFMMfFFhFp",
        "jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL",
        "PmmdzqPrVvPwwTWBwg",
        "wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn",
        "ttgJtRGJQctTZtZT",
        "CrZsJsPPZsGzwwsLwLmpwMDw"
    ]

    func split (_ sack :String) -> (Set<Character>, Set<Character>) {
        (Set(sack.prefix(sack.count/2)), Set(sack.suffix(sack.count/2)))
    }
    func mispacked (_ left :Set<Character>, _ right :Set<Character>) -> Character {
        left.first(where: { right.contains($0) }) ?? "🤔"
    }
    func priority (_ c :Character) -> Int {
        if let v = c.asciiValue { return (Int)(v >= 97 ? v - 96 : (v - 38))  }
        else { return 0 }
    }

    func part1 () throws -> Int { try readInput(3).map(split).map(mispacked).map(priority).reduce(0, +) }

    func badge (_ sacks :[Set<Character>]) -> Character {
        sacks[0].first(where: { c in sacks.suffix(from: 1).allSatisfy({ $0.contains(c) }) }) ?? "🤔"
    }

    func part2 () throws -> Int { 
        try readInput(3).map({ Set($0) }).chunked(by: 3).map(badge).map(priority).reduce(0, +)
    }
}

File: 2022/scala/src/main/scala/Day3.scala
object Day3 extends Day(3):

  def split (sack :String) = (sack.take(sack.size/2).toSet, sack.drop(sack.size/2).toSet)
  def mispacked (left :Set[Char], right :Set[Char]) = (left & right).head
  def priority (c :Char) = if c >= 97 then c - 96 else c - 38
  def badge (sacks :Seq[Set[Char]]) = sacks.reduce(_ & _).head

  override def answer1 (input :Seq[String]) = input.map(split).map(mispacked).map(priority).sum
  override def answer2 (input :Seq[String]) =
    input.map(_.toSet).grouped(3).map(badge).map(priority).sum
